src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+        self.stream.clear_half_transfer_interrupt();
+    }
+
+    /// Clear transfer error interrupt (teif) for the DMA stream.
+    #[inline]
+    pub fn clear_transfer_error_interrupt(&mut self) {
+        self.stream.clear_transfer_error_interrupt();
+    }
+
+    /// Clear direct mode error interrupt (dmeif) for the DMA stream.
+    #[inline]
+    pub fn clear_direct_mode_error_interrupt(&mut self) {
+        self.stream.clear_direct_mode_error_interrupt();
+    }
+
+    /// Clear fifo error interrupt (feif) for the DMA stream.
+    #[inline]
+    pub fn clear_fifo_error_interrupt(&mut self) {
+        self.stream.clear_fifo_error_interrupt();
+    }
+
+    /// Changes the buffer and restarts or continues a double buffer transfer. This must be called
+    /// immediately after a transfer complete event. The closure must return `(BUF, T)` where `BUF`
+    /// is the new buffer to be used. The closure will not be called if the transfer is not
+    /// completed.
+    ///
+    /// # Panics
+    /// This method will panic when double buffering and one or both of the following conditions
+    /// happen:
+    ///
+    /// * The new buffer's length is smaller than the one used in the `init` method.
+    /// * The closure `f` takes too long to return and a buffer overrun occurs.
+    ///
+    /// # Safety
+    /// Memory corruption might occur if the previous buffer, the one passed to the closure, gets
+    /// dropped in the closure and an overrun occurs in double buffering mode.
+    pub unsafe fn next_transfer_with<F, T>(&mut self, f: F) -> Result<T, DMAError<()>>
+    where
+        F: FnOnce(BUF, CurrentBuffer) -> (BUF, T),
+    {
+        if !STREAM::get_transfer_complete_flag() {
+            return Err(DMAError::NotReady(()));
+        }
+        self.stream.clear_transfer_complete_interrupt();
+
+        if self.double_buf.is_some() && DIR::direction() != DmaDirection::MemoryToMemory {
+            let current_buffer = STREAM::current_buffer();
+            // double buffering, unwrap can never fail
+            let db = if current_buffer == CurrentBuffer::DoubleBuffer {
+                self.buf.take().unwrap()
+            } else {
+                self.double_buf.take().unwrap()
+            };
+            let r = f(db, !current_buffer);
+            let mut new_buf = r.0;
+            let (new_buf_ptr, new_buf_len) = new_buf.write_buffer();
+
+            // We can't change the transfer length while double buffering
+            assert!(
+                new_buf_len >= usize::from(self.transfer_length),
+                "Second Buffer not big enough"
+            );
+
+            // If this is false, then RAM corruption might have occurred, there's nothing we can do
+            // apart from panicking.
+            // TODO: Is this the best solution ? The closure based approach seems necessary if we
+            // want to support BBqueue.
+            assert!(current_buffer == STREAM::current_buffer(), "Overrun");
```

Comment:
This check needs to be moved after the `set_memory_address` calls, I think. Only then can we be sure the DMA will not write into the buffer we just gave to the user anymore.